File: files/_semicircles.cpp
#include <cstdio>
#include <algorithm>
#include <utility>
#include <vector>

#define fst first
#define snd second
#define _ make_pair
#define push push_back
#define i(x) scanf("%i",&x)

using namespace std;
const int MOD = 1000000007;

int fillGap[4020];
pair<int, int> circles[1505];
vector<int> under[1505];
int takeAway[1505][4005];
int buffer[4005];

void precompute(int next, int repeat)
{
	if (repeat == 0) return;
	for (int e = 0; e < next; e += 2)
	{
		long long _ = (long long)fillGap[e] * (long long)fillGap[next - 2 - e];
		fillGap[next] = (fillGap[next] + _ % MOD) % MOD;
	}
	precompute(next + 2, repeat - 1);
}
bool sorter(pair<int, int> a, pair<int, int> b){ return a.snd - a.fst < b.snd - b.fst; }
int getWidth(int ind){ return circles[ind].snd - circles[ind].fst + 1; }
void process_(int anon)
{
	//printf("processing node %i\n", anon);
	int width = getWidth(anon);
	if (under[anon].empty()){
		takeAway[anon][0] = 1;
	}
	else{
		for (int e = 0; e < under[anon].size(); process_(under[anon][e++]));
		//map the head
		for (int e_ = 0; e_ <= getWidth(under[anon][0]); e_++)
		{
			takeAway[anon][e_] = takeAway[under[anon][0]][e_];
		}
		int N_SQUARED = getWidth(under[anon][0]);//optimization
		//combinations with the tail
		for (int e_ = 1; e_ < under[anon].size(); e_++)
		{
			for (int e = 0; e < 4003; buffer[e++] = 0);
			for (int e = 0; e <= getWidth(under[anon][e_]); e++)
			{
				//for (int ee = 0; ee <= width; ee++)
				for (int ee = 0; ee <= N_SQUARED;ee++)//optimization
				{
					if (e + ee <= width){
						long long _ = ((long long)takeAway[under[anon][e_]][e] * takeAway[anon][ee]) % MOD;
						buffer[e + ee] = ((long long)buffer[e + ee] + _) % MOD;
					}
				}
			}
			for (int e = 0; e <= width; e++) takeAway[anon][e] = buffer[e];
			//for (int e = 0; e <= width; printf("%i  ", buffer[e++])); printf("\n\n");
			N_SQUARED += getWidth(under[anon][e_]);//optimization
		}
	}
	long long acc = 0;
	for (int e = 0; e <= width; e++)
	{
		if (width - e - 2 >= 0){
			long long _ = (long long)takeAway[anon][e] * fillGap[width - e - 2];
			acc += _ % MOD;
		}
	}
	takeAway[anon][width] = acc % MOD;
	//for (int e = 0; e <= width; printf("%i  ", takeAway[anon][e++])); printf("\n\n");
}

int main()
{
	int n; i(n);
	int k; i(k);
	for (int e = 0; e < k; e++)
	{
		int a, b; i(a); i(b);
		circles[e] = _(a - 1, b - 1);
	}
	//precompute catalan numbers
	fillGap[0] = 1;
	precompute(2, n / 2 + 5);

	//tree
	sort(circles, circles + k, sorter);
	circles[k] = _(-1, n);
	//like a normal person
	for (int e_ = 0; e_ <= k; e_++)
	{
		for (int ee_ = e_ + 1; ee_ <= k; ee_++)
		{
			pair<int, int> e = circles[e_];
			pair<int, int> ee = circles[ee_];
			if (ee.fst<e.fst && ee.snd>e.snd){
				under[ee_].push(e_);
				break;
			}
		}
	}
	process_(k);
	printf("%i", takeAway[k][n + 2]);

	scanf(" ");
	return 0;
}

File: files/_semicircles.py
import sys
def i(): return sys.stdin.readline().strip()
sys.setrecursionlimit(99999999)
MOD=1000000007

n=int(i())
k=int(i())
circles=[map(int,i().split(" ")) for e in xrange(k)]
#precompute
def precompute(ll,repeat):
    if repeat==0: return ll
    else:
        backward=ll[:]
        backward.reverse()
        ll.append(sum(map(lambda a:(a[0]*a[1])%MOD,zip(ll,backward)))%MOD)
        return precompute(ll,repeat-1)
fillGap=[]
for e in precompute([1,1],n/2+5):
    fillGap.append(e)
    fillGap.append(0)

#establish tree
circles.sort()
circles.reverse()
tree=[(e[0]-1,e[1]-1) for e in circles]+[(-1,n)]
#determine if the next element in the tree list is NOT part of the current element's subtree
def stump(e,anon):
    if len(anon)==0: return True
    else:
        return e[1]<anon[-1][0]
#traverse tree (make it into an actual tree)
def traverse(ll):
    if len(ll)==0: return (0,[])
    else:
        hd=ll[-1]
        ll.pop()
        nodes=[]
        while not stump(hd,ll):
            nodes.append(traverse(ll))
        return (hd[1]-hd[0]+1,nodes)
    
#process tree
def process_(anon):
    width,ll=anon
    #take away i dots, to create a need to fill in width-i dots using precomputed table
    takeAway=[0]*(width+1)
    if len(ll)==0: takeAway[0]=1
    #combine all retrieved arrays (dfs)
    else:
        retrieved=map(process_,ll)
        #map the head onto the thing
        for e in xrange(len(retrieved[-1])):
            takeAway[e]+=retrieved[-1][e]
        N_SQUARED=len(retrieved[-1])-1#optimization

        #combinations with the tail
        retrieved.pop()
        for e_ in retrieved:
            nn=[0]*len(takeAway)        #buffer
            for e in xrange(len(e_)):
                #for ee in xrange(len(nn)):
                for ee in xrange(N_SQUARED+1):#optimization
                    if e+ee<len(nn):
                        nn[e+ee]=(nn[e+ee]+e_[e]*takeAway[ee])%MOD
            takeAway=nn[:]
            N_SQUARED+=len(e_)-1#optimization
    #if the semicircle representing this node was not erased:
    #use precomputed table on each amount to find out num. ways
    acc=0
    for e in xrange(len(takeAway)):
        if width-e-2>=0:
            acc+=(takeAway[e]*fillGap[width-e-2])%MOD
    takeAway[width]=acc%MOD
    return takeAway
print process_(traverse(tree))[-1]

raw_input()
